src/compiler/checker.rs

```diff
@@ -1148,20 +1220,31 @@ impl TypeChecker {
     ) -> Rc<Type> {
         let union_reduction = union_reduction.unwrap_or(UnionReduction::Literal);
         if types.is_empty() {
-            unimplemented!()
-            // return self.never_type();
+            return self.never_type();
         }
         if types.len() == 1 {
             return types[0].clone();
         }
         let mut type_set: Vec<Rc<Type>> = vec![];
-        // let includes =
-        self.add_types_to_union(&mut type_set, /*TypeFlags::empty(), */ &types);
+        let includes = self.add_types_to_union(&mut type_set, TypeFlags::None, &types);
         if union_reduction != UnionReduction::None {}
-        self.get_union_type_from_sorted_list(type_set)
+        let object_flags = (if includes.intersects(TypeFlags::NotPrimitiveUnion) {
+            ObjectFlags::None
+        } else {
+            ObjectFlags::PrimitiveUnion
+        }) | (if includes.intersects(TypeFlags::Intersection) {
+            ObjectFlags::ContainsIntersections
+        } else {
+            ObjectFlags::None
+        });
```

Comment:
I find the if-statements inside parentheses, or-ed together pretty hard to parse here.